File: 04-stateful-stream-processing/java-scala/e2/src/main/java/com/becomedataengineer/DataGenerator.java
package com.becomedataengineer;

import com.fasterxml.jackson.core.JsonProcessingException;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;

import java.io.IOException;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.Properties;
import java.util.concurrent.TimeUnit;

import static com.becomedataengineer.KafkaConfiguration.ALL_TOPICS;

public class DataGenerator {

    public static void main(String[] args) throws Exception {
        System.out.println(">>> Creating topics...");
        ALL_TOPICS.forEach(topicName -> {
            try {
                Process deleteTopicCommand = null;
                deleteTopicCommand = Runtime.getRuntime()
                        .exec("docker exec e2_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic "+topicName +" --delete");
                deleteTopicCommand.waitFor();
                deleteTopicCommand.exitValue();
                Process createTopicCommand = null;
                createTopicCommand = Runtime.getRuntime()
                        .exec("docker exec e2_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic "+topicName +" --create --partitions 2");
                createTopicCommand.waitFor();
                if (createTopicCommand.exitValue() != 0) {
                    throw new RuntimeException("Topic " + topicName + " couldn't be created");
                }
            } catch (IOException | InterruptedException e) {
                throw new RuntimeException(e);
            }
        });
        System.setProperty("user.timezone", "UTC");
        Properties producerProperties = new Properties();
        producerProperties.setProperty("bootstrap.servers", "localhost:29092");
        producerProperties.setProperty("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
        producerProperties.setProperty("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");
        KafkaProducer<String, String> producer = new KafkaProducer<>(producerProperties);
        ZonedDateTime eventTime = ZonedDateTime.now(ZoneId.of("UTC"));
        int toGenerate = 1000000;
        while (toGenerate-- > 0) {
            System.out.println(">>> Generating records");
            eventTime = eventTime.plusSeconds(30);
            try {
                String jsonA = Json.MAPPER.writeValueAsString(new Visit("1", eventTime));
                producer.send(new ProducerRecord<>(KafkaConfiguration.INPUT_TOPIC_NAME,  "1", jsonA));
                String jsonB = Json.MAPPER.writeValueAsString(new Visit("2", eventTime));
                producer.send(new ProducerRecord<>(KafkaConfiguration.INPUT_TOPIC_NAME,  "2", jsonB));
                String jsonC = Json.MAPPER.writeValueAsString(new Visit("3", eventTime));
                producer.send(new ProducerRecord<>(KafkaConfiguration.INPUT_TOPIC_NAME,  "3", jsonC));
                producer.flush();
            } catch (JsonProcessingException e) {
                throw new RuntimeException(e);
            }

            Thread.sleep(TimeUnit.SECONDS.toMillis(2));
        }
    }

}

File: 04-stateful-stream-processing/python/e2/data_generator.py
import json
import os
import random
import time
from datetime import datetime, timedelta

from confluent_kafka import Producer

from config import get_input_topic_name, get_output_topic_name

print('>>> Recreating topics...')
for topic_name in [get_input_topic_name(), get_output_topic_name()]:
    print(f'....{topic_name}')
    os.system(
        f'docker exec e2_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic {topic_name} --delete')
    os.system(
        f'docker exec e2_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic {topic_name} --create --partitions 2')


def visit(visit_id: str, event_time: datetime) -> str:
    return json.dumps({'visit_id': visit_id,
                       'event_time': event_time.isoformat(),
                       'browser': random.choice(["Firefox", "Chrome", "Safari"])})


event_time_for_events = datetime.utcnow()
producer = Producer({'bootstrap.servers': 'localhost:29092'})
while True:
    event_time_for_events += timedelta(seconds=30)
    print('>>>> Generating records')
    producer.produce(topic=get_input_topic_name(), partition=0, value=visit('1', event_time_for_events))
    producer.produce(topic=get_input_topic_name(), partition=1, value=visit('2', event_time_for_events))
    producer.produce(topic=get_input_topic_name(), partition=1, value=visit('3', event_time_for_events))
    producer.flush()

    time.sleep(5)
